pkg/api/validate/dynamic/dynamic.go

```diff
@@ -617,34 +623,92 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
 			)
 		}
 
-		if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
-			if ss.SubnetPropertiesFormat != nil &&
-				ss.SubnetPropertiesFormat.NetworkSecurityGroup != nil {
-				expectedNsgID, err := subnet.NetworkSecurityGroupID(oc, s.ID)
+		subnetByID[s.ID] = ss
+	}
+	return subnetByID, nil
+}
+
+// checkByoNSG checks whether all the subnets have or don't have NSG attached.
+// when the BYONsg feature flag is on and only some of the subnets are attached with an NSG,
+// it returns an error.  If none of the subnets is attached, it's no longer BYO NSG and the
+// cluster installation process should fall back to using the managed nsg.
+func (dv *dynamic) checkByoNSG(subnetByID map[string]*mgmtnetwork.Subnet) (api.PreconfiguredNSG, error) {
```

Comment:
Looking back, it's not clear to me why we need `checkByoNSG()` at all (can we just have this logic in ValidateSubnets)?

The validation logic in the PR is split between `checkByoNSG()` and `ValidateSubnets()`, where the first one determines if subnets have / don't have NSGs attached, adjusting the doc if needed; and the second one takes it a step further and makes sure it's the same NSG.

Do we want them in separate functions because one is making changes to the cluster document, and the other is just validating? Or can we combine logic a little?

Reply:
I'm +1 on the `byo` vs `preconfigured` language - I'd rather pick one and go with it. If the BU-approved AFEC is `BYO`, let's use that everywhere so we have a 1:1 correlation of names to features.

Reply:
I have changed all to `preconfigured` instead of `byo`.  

About the validation logic, `checkPreconfiguredNSG` (renamed) has to happen first before the existing validation in `ValidateSubnets()`.  This is because depending on the result of `checkPreconfiguredNSG`, it could turn the feature off and the validation logic then must change accordingly.  Having it in a separate function makes it easier to write unit tests for its test scenarios.  Also, `checkPreconfiguredNSG` most likely will be removed once the feature is GAed.  Once the real API change is introduced, the validation will be different (e.g. also rejecting when no NSG is attached at all).  So I think we should not combine them, at least for now.

Reply:
Ok - thanks for the explanation.